### PR TITLE
Fixing translation request batching

### LINKED ISSUE #18: Error - The maximum request size has been exceeded.
We are in the process of using Azure Cognitive Services for translating a rather large enterprise application. When attempting to translate our resource files to multiple languages the following error is raised. The configured tier in Azure is **S0**. The error message appears to be triggered on longer sentences in our **resx** files. It's impossible with the current error message to identify which resource file and key that is causing the problem. 

image

### PR DESCRIPTION
**Pull request by vs-savelich**
- Fixed locales batching.
- Implemented request batching when file is too large (more than 10 000 characters or more than 100 resource keys).

Fixes #18

### CODE REVIEW

File: __tests__/api.test.ts [hunk 1]
```diff
@@ -39,15 +69,16 @@ test('API: get available translations correctly gets all locales', async () => {
     expect(locales).toEqual(expectedLocales.join(', '));
 });
 
-test('API: read file->translate->apply->write', async () => {
+test.skip('API: read file->translate->apply->write', async () => {
```
Comment (IEvangelist) [inline __tests__/api.test.ts:72]: Why is this skipped? Did something break with the read, translate, apply write?

File: __tests__/api.test.ts [hunk 2]
```diff
@@ -90,11 +122,11 @@ test('API: read file->translate->apply->write', async () => {
     }
 
     const [title, details] = summarize(summary);
-    expect(title).toEqual('Machine-translated 79 files, a total of 1,185 translations');
+    expect(title).toEqual('Machine-translated 89 files, a total of 1,335 translations');
     expect(details).toBeTruthy();
 });
 
-test('API: translate correctly performs translation', async () => {
+test.skip('API: translate correctly performs translation', async () => {
```
Comment (IEvangelist) [inline __tests__/api.test.ts:129]: Same question here, why is this being skipped now?

File: src/api/translation-api.ts [hunk 3]
```diff
@@ -1,28 +1,46 @@
...
     return response.data;
-}
+};
 
 export const translate = async (
     translatorResource: TranslatorResource,
     toLocales: string[],
-    translatableText: Map<string, string>): Promise<TranslationResultSet | undefined> => {
+    translatableText: Map<string, string>,
+    filePath: string): Promise<TranslationResultSet | undefined> => {
```
Comment (IEvangelist) [inline src/api/translation-api.ts:25]: What is the `filePath` being used for, it is hard to tell?

File: src/api/translation-api.ts [hunk 4]
```diff
@@ -1,28 +1,46 @@
...
+        const numberOfElementsLimit = 100;
+
+        const validationErrors: string[] = [];
+        translatableText.forEach((value, key) => {
+            const valueStringifiedLength = JSON.stringify(value).length;
+            if (valueStringifiedLength > apiRateLimit) {
+                validationErrors.push(`Text for key '${key}' in file '${filePath}' is too long (${valueStringifiedLength}). Must be ${apiRateLimit} at most.`);
+            }
+        });
+        if (validationErrors.length) {
```
Comment (IEvangelist) [inline src/api/translation-api.ts:39]: Is this a new early exit? If so, do you anticipate it ever actually happening? Is there anyway to avoid that, and somehow buffer partial calls? That would probably be a separate PR - just thinking out loud.

File: src/abstractions/translation-file-kind.ts [hunk 5]
```diff
@@ -23,7 +23,20 @@ export type TranslationFileKind =
...
+     * Supports nested objects, e.g.
+     * {
+     *      "messages": {
+     *          "foo": {
+     *              "bar": "Hello!"
+     *          }
+     *      }
+     * }
+    */
+    'json';
```
Comment (IEvangelist) [inline src/abstractions/translation-file-kind.ts:42]: This is a great addition, thank you!

File: __tests__/api.test.ts [hunk 6]
```diff
@@ -90,11 +122,11 @@ test('API: read file->translate->apply->write', async () => {
     }
 
     const [title, details] = summarize(summary);
-    expect(title).toEqual('Machine-translated 79 files, a total of 1,185 translations');
+    expect(title).toEqual('Machine-translated 89 files, a total of 1,335 translations');
```
Comment (IEvangelist) [inline __tests__/api.test.ts:125]: Yeah, this test -- ugh 😬. Thank you

File: README.md [hunk 7]
```diff
@@ -27,6 +28,7 @@ on:
     - '**.restext'   # INI-based, key value pair file format
     - '**.resx'      # XML-based (resource) translation file format, .NET
     - '**.xliff'     # XML-based translation file format, version 2
+    - '**.json'      # JSON-based, key value pair file format
```
Comment (IEvangelist) [inline README.md:31]: Wow, this is great - is there anyway you could update the PR to include updates to the README.md file as well that add .json support?

**Review by IEvangelist — COMMENTED** [review]
Once I hear back on these asks, I'd be thrilled to approve and merge - thank you so much! This is awesome.

File: __tests__/api.test.ts (hunk 1 above)
Comment (vs-savelich) [reply __tests__/api.test.ts:72]: These tests were initially skipped, probably because they don't mock real Azure call. There was a dummy test "test.only('API: tests are currently ignored - comment out this test to run actual tests.')". I can't use same technic to ignore some tests, because I added some real tests to this file.

File: __tests__/api.test.ts (hunk 2 above)
Comment (vs-savelich) [reply __tests__/api.test.ts:129]: See previous comment 😉

File: src/api/translation-api.ts (hunk 3 above)
Comment (vs-savelich) [reply src/api/translation-api.ts:25]: It used only to improve error messages. Previously it was impossible to know which file causes a problem.

File: src/api/translation-api.ts (hunk 4 above)
Comment (vs-savelich) [reply src/api/translation-api.ts:39]: It can happen if one have more than 10000 chars in a single resource text. I believe it is better to fail fast here without calling translation service.
Yes, we can do partial calls, but it would require to implement even more sophisticated buffering strategy here. In other words, we need to know how to split single resource text. Perhaps by '.' char? I believe that letting users know that they have some too long text, is good enough solution for now.

**Review by IEvangelist — APPROVED** [review]

**Comment by IEvangelist** [comment]
Hi @vs-savelich and @peters 

This is awesome, thank you very much for your contributions - it's packaged and published to the marketplace as [v2.1.4](https://github.com/IEvangelist/resource-translator/releases/tag/v2.1.4). Please feel free to report any new bugs, or issues - thanks again.

**Comment by IEvangelist** [comment]
@all-contributors please add @vs-savelich for code and tests
